### ISSUE TITLE
机器学习的开始

### ISSUE THREAD

**Issue by Kamazuki** [Kamazuki/blogs#1]
机器学习框架非常多。挑了一个比较容易上手的 keras 来学习。

### 一、资料准备：

1. Keras 中文文档 https://keras.io/zh/
2. TensorFlow 中文文档 https://www.tensorflow.org/tutorials/quickstart/beginner?hl=zh_cn
3. StackOverFlow 😂

### 二、环境准备：

本来想搭一台云主机来进行机器学习的，因为感觉本地机器运行的性能可能很慢。
腾讯云、阿里云看了很久，准备下单的腾讯云的GPU云服务器的时候，想到说不定过会太难就不学习了。金钱的罪恶感使我踏上了免费资源寻找之路。
没想到这么简单就找到了。

Google 的 colab https://colab.research.google.com/
免费，还帮我搭好了环境，太省心了。

### 三、hello world

看了n篇文章，发现手写数字识别就是机器学习（神经网络）的 “hello world” 程序了，目标就是你！

代码如下：

    import keras
    
    #Sequential : 模型的一种，采用顺序添加神经网络层的一种模型组织形式的类
    from keras.models import Sequential
    
    #Dense, Activation, Dropout : 神经网络的层，用于定义层的输入输出
    from keras.layers import Dense, Activation, Dropout

    #mnist : keras 提供的预定义数据
    from keras.datasets import mnist

    #RMSprop : keras预定义的一种优化器
    from keras.optimizers import RMSprop

    #这里有多个返回值
    (x_train, y_train), (x_test, y_test) = mnist.load_data()

    #原始加载出来的x_train\x_test 是 numpy.ndarray 的类型，由 28 * 28 的 uint8 矩阵组成？
    #需要把它们转化成一维的数据（后面定义的神经网络层只支持一维数据）
    x_train = x_train.reshape(60000, 28 * 28)
    x_test = x_test.reshape(10000, 28 * 28)

    #默认数据是 uint8 类型[0, 255]，后面要归一化，所以先转化成 float32 类型
    x_train = x_train.astype('float32')
    x_test = x_test.astype('float32')

    #这里的 x_train 实际上是 numpy.ndarray 类型，重载了一些运算符，除以255 进行归一化 
    x_train /= 255
    x_test /= 255

    #y_train\y_test 是 numpy.ndarray 类型，由 uint8 组成
    #因为我们需要做的是分类器（将输入的图片分类成 0～9 个分类）
    #因此将数据从 uint8 转化成长度为10的数组，转化后 result[i] = (uint8)(y_train == i)
    y_train = keras.utils.to_categorical(y_train, 10)
    y_test = keras.utils.to_categorical(y_test, 10)

    model = Sequential()

    #28*28 == 784, 这里是定义一个接受784个输入的神经网络层（对应一张图片），512个输出
    model.add(Dense(512, input_shape=(784,)))

    #激活层
    model.add(Activation('relu'))

    #随机丢弃部分数据，避免过拟合
    model.add(Dropout(0.2))

    #两层神经网络层之间，会自动匹配输入，所以这里只需要定义512个输出
    model.add(Dense(512))

    #激活层
    model.add(Activation('relu'))

    #随机丢弃部分数据，避免过拟合
    model.add(Dropout(0.2))

    #这里是最后一层，定义了10个输出，对应10个分类
    model.add(Dense(10))

    #使用了softmax 的激活函数来进行归一化，输出的数据范围[0, 1], 正好对应每一个分类的概率
    model.add(Activation('softmax'))

    #keres 预定义的优化函数
    opt = RMSprop()
 
    #编译模型
    model.compile(loss='mse', optimizer=opt, metrics=['accuracy'])

    #进行模型训练，20次迭代
    model.fit(x_train, y_train, batch_size=128, epochs=20)

    #根据测试集来进行打分
    score = model.evaluate(x_test, y_test)

    #输出最后模型的得分
    print('Total loss on Testing Set:', score[0])
    print('Accuracy of Testing Set', score[1])

colab 里面运行就能看到以下的输出：

    Using TensorFlow backend.
    Downloading data from https://s3.amazonaws.com/img-datasets/mnist.npz
    11493376/11490434 [==============================] - 0s 0us/step
    Epoch 1/20
    60000/60000 [==============================] - 8s 135us/step - loss: 0.0123 - accuracy: 0.9166
    Epoch 2/20
    60000/60000 [==============================] - 8s 133us/step - loss: 0.0057 - accuracy: 0.9634
    Epoch 3/20
    60000/60000 [==============================] - 8s 135us/step - loss: 0.0043 - accuracy: 0.9726
    Epoch 4/20
    60000/60000 [==============================] - 8s 133us/step - loss: 0.0035 - accuracy: 0.9775
    Epoch 5/20
    60000/60000 [==============================] - 8s 132us/step - loss: 0.0032 - accuracy: 0.9796
    Epoch 6/20
    60000/60000 [==============================] - 8s 133us/step - loss: 0.0027 - accuracy: 0.9830
    Epoch 7/20
    60000/60000 [==============================] - 8s 134us/step - loss: 0.0025 - accuracy: 0.9843
    Epoch 8/20
    60000/60000 [==============================] - 8s 134us/step - loss: 0.0022 - accuracy: 0.9861
    Epoch 9/20
    60000/60000 [==============================] - 8s 135us/step - loss: 0.0021 - accuracy: 0.9869
    Epoch 10/20
    60000/60000 [==============================] - 8s 133us/step - loss: 0.0020 - accuracy: 0.9875
    Epoch 11/20
    60000/60000 [==============================] - 8s 132us/step - loss: 0.0018 - accuracy: 0.9886
    Epoch 12/20
    60000/60000 [==============================] - 8s 132us/step - loss: 0.0017 - accuracy: 0.9897
    Epoch 13/20
    60000/60000 [==============================] - 8s 132us/step - loss: 0.0016 - accuracy: 0.9897
    Epoch 14/20
    60000/60000 [==============================] - 8s 134us/step - loss: 0.0016 - accuracy: 0.9904
    Epoch 15/20
    60000/60000 [==============================] - 8s 131us/step - loss: 0.0013 - accuracy: 0.9916
    Epoch 16/20
    60000/60000 [==============================] - 8s 132us/step - loss: 0.0013 - accuracy: 0.9922
    Epoch 17/20
    60000/60000 [==============================] - 8s 131us/step - loss: 0.0014 - accuracy: 0.9916
    Epoch 18/20
    60000/60000 [==============================] - 8s 132us/step - loss: 0.0012 - accuracy: 0.9924
    Epoch 19/20
    60000/60000 [==============================] - 8s 131us/step - loss: 0.0012 - accuracy: 0.9925
    Epoch 20/20
    60000/60000 [==============================] - 8s 132us/step - loss: 0.0012 - accuracy: 0.9927
    10000/10000 [==============================] - 1s 70us/step
    Total loss on Testing Set: 0.0028493214674062887
    Accuracy of Testing Set 0.9829000234603882

恭喜你，你的机器学习 hello world 已经成功！😊

### 四、保存数据

为了能在 colab 上面愉快玩耍，需要将 colab 跟 Google Drive 的路径绑定起来，这样就可以把训练的数据保存到 Google Drive 上面。

    from google.colab import drive

    #这里是绑定，注意路径不要打错，这个路径是固定的。
    rootPath = '/content/drive/'
    drive.mount(rootPath, force_remount=True)

    #Google Drive 的云端硬盘目录是 My Drive，这个也不要打错，后面的目录或者文件名字就可以自定义了
    path = rootPath + 'My Drive/firstDemo.h5'

    #采用写+创建的方式打开文件，避免文件不存在导致无法保存
    f = open(path, 'w+')
    f.close()

    #保存目录
    model.save(path)

colab 会看到以下输出：

    Go to this URL in a browser: https://accounts.google.com/o/oauth2/XXX

    Enter your authorization code:
    ··········
    Mounted at /content/drive/

这里需要注意的是，保存数据到 Google Drive 需要 oauth 授权，点击 https://accounts.google.com/o/oauth2/XXX 这个路径，就会跳转到授权页面，授权完之后复制授权码，回到 colab，填到 Enter your authorization code: 下方的输入框，然后点击 Enter 即可完成授权。

这样就能在Google Drive 的 “我的云端硬盘” 里面看到刚刚保存的 firstDemo.h5 文件。

### What's next？

下一篇，让我们将训练好的模型运用到 iOS 机器上吧。
